EKF/control.cpp

```diff
@@ -757,10 +757,19 @@ void Ekf::controlHeightFusion()
 		// we have just switched to using range finder, calculate height sensor offset such that current
 		// measurment matches our current height estimate
 		if (_control_status_prev.flags.rng_hgt != _control_status.flags.rng_hgt) {
-			if (_terrain_initialised) {
-				_hgt_sensor_offset = _terrain_vpos;
+			// use the parameter rng_gnd_clearance if on ground to avoid a noisy offset initialization (e.g. sonar)
+			if (_control_status.flags.in_air) {
+
+				if (_terrain_initialised) {
+
+					_hgt_sensor_offset = _terrain_vpos;
+				} else {
+
+					_hgt_sensor_offset = _R_rng_to_earth_2_2 * _range_sample_delayed.rng + _state.pos(2);
```

Comment:
There is one case where it would not behave correctly, I think. If you start with baro at a certain altitude and then go e.g. uphill and then switch to range sensor and the terrain is not valid (`_terrain_initialized` should/will be replaced by `get_terrain_vert_pos()`, see https://github.com/PX4/ecl/pull/301), it will use `_hgt_sensor_offset = _params.rng_gnd_clearance;` and this has the effect of a sudden height change of the drone which is undesired.